sys/userns_unsafe_linux.go

```diff
@@ -22,44 +22,73 @@ import (
 	"unsafe"
 )
 
-// ProcSyncType is used for synchronization
-// between parent and child processes.
-type ProcSyncType uint8
-
-const (
-	// ProcSyncExit tells child "it's time to exit".
-	ProcSyncExit ProcSyncType = 0x1
-)
-
+// ForkUserns is to fork child process with user namespace. It returns child
+// process's pid and pidfd reference to the child process.
+//
+// Precondition: The runtime OS thread must be locked, which is GO runtime
+// requirement.
+//
+// Beside this, the child process sets PR_SET_PDEATHSIG with SIGKILL so that
+// the parent process's OS thread must be locked. Otherwise, the exit event of
+// parent process's OS thread will send kill signal to the child process,
+// even if parent process is still running.
+//
 //go:norace
 //go:noinline
-func ForkUserns(pipeMap [2]int) (pid uintptr, errno syscall.Errno) {
-	var sync ProcSyncType
+func ForkUserns() (_pid uintptr, _pidfd uintptr, _ syscall.Errno) {
+	var (
+		pidfd     uintptr
+		pid, ppid uintptr
+		err       syscall.Errno
+	)
+
+	ppid, _, err = syscall.RawSyscall(uintptr(syscall.SYS_GETPID), 0, 0, 0)
+	if err != 0 {
+		return 0, 0, err
+	}
 
 	beforeFork()
 	if runtime.GOARCH == "s390x" {
-		pid, _, errno = syscall.RawSyscall6(uintptr(syscall.SYS_CLONE), 0, syscall.CLONE_NEWUSER|uintptr(syscall.SIGCHLD), 0, 0, 0, 0)
+		// NOTE:
+		//
+		// On the s390 architectures, the order of the first two
+		// arguments is reversed.
+		//
+		// REF: https://man7.org/linux/man-pages/man2/clone.2.html
+		pid, _, err = syscall.RawSyscall(syscall.SYS_CLONE,
+			0,
+			uintptr(syscall.CLONE_NEWUSER|syscall.SIGCHLD|syscall.CLONE_PIDFD),
+			uintptr(unsafe.Pointer(&pidfd)),
+		)
 	} else {
-		pid, _, errno = syscall.RawSyscall6(uintptr(syscall.SYS_CLONE), syscall.CLONE_NEWUSER|uintptr(syscall.SIGCHLD), 0, 0, 0, 0, 0)
+		pid, _, err = syscall.RawSyscall(syscall.SYS_CLONE,
+			uintptr(syscall.CLONE_NEWUSER|syscall.SIGCHLD|syscall.CLONE_PIDFD),
+			0,
+			uintptr(unsafe.Pointer(&pidfd)),
+		)
 	}
-	if errno != 0 || pid != 0 {
+	if err != 0 || pid != 0 {
 		afterFork()
-		return pid, errno
+		return pid, pidfd, err
 	}
-
 	afterForkInChild()
-	if _, _, errno = syscall.RawSyscall(syscall.SYS_CLOSE, uintptr(pipeMap[1]), 0, 0); errno != 0 {
+
+	if _, _, err = syscall.RawSyscall(syscall.SYS_PRCTL, syscall.PR_SET_PDEATHSIG, uintptr(syscall.SIGKILL), 0); err != 0 {
 		goto err
 	}
-	if _, _, errno = syscall.RawSyscall6(syscall.SYS_PRCTL, syscall.PR_SET_PDEATHSIG, uintptr(syscall.SIGKILL), 0, 0, 0, 0); errno != 0 {
+
+	pid, _, err = syscall.RawSyscall(syscall.SYS_GETPPID, 0, 0, 0)
+	if err != 0 {
 		goto err
 	}
-	// wait for parent's signal
-	if _, _, errno = syscall.RawSyscall6(syscall.SYS_READ, uintptr(pipeMap[0]), uintptr(unsafe.Pointer(&sync)), unsafe.Sizeof(sync), 0, 0, 0); errno != 0 || sync != ProcSyncExit {
+
+	// exit if re-parent
+	if pid != ppid {
 		goto err
 	}
 
+	_, _, err = syscall.RawSyscall(syscall.SYS_PPOLL, 0, 0, 0)
```

Comment:
I didn't see sys_pause syscall in riscv64 so I use ppoll on empty fd to make the thread into infinite sleep status